File: modules/pipeline.py
"""Resumable generation pipeline: configuration, locking, scheduling, stitching."""

import concurrent.futures
import fcntl
import hashlib
import json
import os
import shutil
import tempfile
from dataclasses import dataclass
from pathlib import Path

import webvtt

from modules import core, gemini, io, media

CHUNK_ROOT = "temp_video_chunks"
LOCK_NAME = ".lock"
DEFAULT_API_WORKERS = 7


@dataclass(frozen=True, slots=True)
class GenerationConfig:
    """Immutable inputs for one generation run."""

    video_path: Path
    output_path: Path
    model: str
    api_key: str | None = None
    base_url: str | None = None
    refine_model: str | None = None
    audio_refine_model: str | None = None
    chunk_dur: int = 60
    workers: int = DEFAULT_API_WORKERS
    thinking_level: str | None = None
    audio_refine: bool = True
    refine_text: bool = True
    context_urls: tuple[str, ...] = ()

    @property
    def chunk_thinking_level(self) -> str:
        """Resolved chunk thinking level used by the manifest and API calls."""
        return self.thinking_level or gemini.DEFAULT_CHUNK_THINKING_LEVEL


def build_manifest(config: GenerationConfig):
    """Build the manifest dictionary and resolved work directory path."""
    ext, mime, video_codec = media.probe_video_format(str(config.video_path))

    manifest = {
        "video": io.file_fingerprint(config.video_path),
        "chunk_dur": config.chunk_dur,
        "format": "stream-copy-v1",
        "mode": "generate",
        "model": config.model,
        "chunk_thinking_level": config.chunk_thinking_level,
        "chunk_ext": ext,
        "chunk_mime": mime,
        "video_codec": video_codec,
    }
    digest = hashlib.sha256(
        json.dumps(manifest, sort_keys=True).encode("utf-8")
    ).hexdigest()[:16]
    return manifest, os.path.join(CHUNK_ROOT, digest)


def acquire_lock(chunk_dir):
    """Acquire an exclusive advisory POSIX lock on the work directory."""
    lock_path = os.path.join(chunk_dir, LOCK_NAME)
    lock_file = open(  # noqa: SIM115 - The caller keeps the lock file open.
        lock_path, "a+", encoding="utf-8"
    )
    try:
        fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
    except BlockingIOError:
        lock_file.seek(0)
        pid = lock_file.read().strip()
        lock_file.close()
        detail = f" (PID {pid})" if pid.isdigit() else ""
        raise RuntimeError(
            f"Another run{detail} is already using {chunk_dir}"
        ) from None

    lock_file.seek(0)
    lock_file.truncate()
    lock_file.write(str(os.getpid()))
    lock_file.flush()
    return lock_file


def release_lock(lock_file):
    """Release the advisory lock and close the lock file descriptor."""
    if lock_file:
        try:
            fcntl.flock(lock_file.fileno(), fcntl.LOCK_UN)
        finally:
            lock_file.close()


def clean_completed_work(chunk_dir):
    """Remove all work directory contents except the active lock file."""
    for entry in os.scandir(chunk_dir):
        if entry.name == LOCK_NAME:
            continue
        if entry.is_dir(follow_symlinks=False):
            shutil.rmtree(entry.path)
        else:
            os.remove(entry.path)


def collect_api_results(futures):
    """Collect completed futures and return names of failed chunks."""
    failed = []
    for future in concurrent.futures.as_completed(futures):
        chunk_name = futures[future]
        try:
            if not future.result():
                failed.append(chunk_name)
        except Exception as e:  # noqa: BLE001 - Convert worker failures to results.
            print(f"ERROR processing {chunk_name}: {e}")
            failed.append(chunk_name)
    return failed


def process_chunks(
    api_key,
    base_url,
    chunk_dir,
    chunks,
    api_workers,
    model_name,
    chunk_mime,
    thinking_level,
    source_title=None,
):
    """Process video chunks concurrently with a thread pool and collect failures."""
    print(f"Processing {len(chunks)} chunks using {api_workers} workers...")
    with concurrent.futures.ThreadPoolExecutor(max_workers=api_workers) as executor:
        futures = {
            executor.submit(
                gemini.process_chunk,
                api_key,
                base_url,
                chunk,
                chunk_dir,
                model_name,
                chunk_mime,
                thinking_level,
                source_title,
            ): chunk["name"]
            for chunk in chunks
        }
        return collect_api_results(futures)


def stitch(chunk_dir, output_vtt):
    """Stitch stream-copy chunk results into one VTT at segment offsets."""
    print("Stitching chunks into final VTT...")
    chunks = media.list_chunks(chunk_dir)
    boundary_starts = [chunk["start"] for chunk in chunks[1:]]
    chunk_by_index = {chunk["idx"]: chunk for chunk in chunks}

    json_files = sorted(
        [
            f
            for f in os.listdir(chunk_dir)
            if f.startswith("subtitle_chunk_") and f.endswith(".json")
        ]
    )
    result_indices = {
        int(name.removeprefix("subtitle_chunk_").removesuffix(".json"))
        for name in json_files
    }
    expected_indices = set(chunk_by_index)
    missing_indices = sorted(expected_indices - result_indices)
    unexpected_indices = sorted(result_indices - expected_indices)
    if missing_indices or unexpected_indices:
        problems = []
        if missing_indices:
            problems.append(f"missing chunk indices: {missing_indices}")
        if unexpected_indices:
            problems.append(f"unexpected chunk indices: {unexpected_indices}")
        raise ValueError(f"Invalid subtitle results: {'; '.join(problems)}")

    entries = []
    for json_name in json_files:
        chunk_idx = int(json_name.removeprefix("subtitle_chunk_").removesuffix(".json"))
        offset_sec = chunk_by_index[chunk_idx]["start"]

        with open(os.path.join(chunk_dir, json_name), "r", encoding="utf-8") as f:
            captions = json.load(f)

        for cap in captions:
            abs_start = offset_sec + core.parse_time(cap["start"])
            abs_end = offset_sec + core.parse_time(cap["end"])
            if abs_end <= abs_start:
                raise ValueError(f"Invalid caption timing in {json_name}: {cap}")

            entries.append(
                {
                    "start": abs_start,
                    "end": abs_end,
                    "text": cap["text"],
                    "chunk_idx": chunk_idx,
                }
            )

    entries.sort(key=lambda item: item["start"])
    entries = core.merge_visual_boundary_fragments(entries, boundary_starts)

    final_vtt = webvtt.WebVTT()
    for entry in entries:
        final_vtt.captions.append(
            webvtt.Caption(
                core.format_time(entry["start"]),
                core.format_time(entry["end"]),
                entry["text"],
            )
        )

    io.atomic_save_vtt(final_vtt, output_vtt)
    print(
        f"Successfully saved to {output_vtt} with {len(final_vtt.captions)} total captions."
    )
    return output_vtt


def validate_generation_config(config: GenerationConfig) -> list[str]:
    """Validate pipeline generation configuration inputs."""
    context_urls = core.validate_context_urls(config.context_urls)

    if config.chunk_dur <= 0:
        raise ValueError("--chunk-dur must be greater than 0")

    if config.workers <= 0:
        raise ValueError("--workers must be greater than 0")

    gemini.validate_thinking_level_for_model(config.model, config.chunk_thinking_level)

    if not config.video_path.exists():
        raise RuntimeError(f"Video file not found: {config.video_path}")

    if config.video_path.resolve() == config.output_path.resolve():
        raise RuntimeError("--output must not resolve to the source video")

    if not config.api_key:
        raise RuntimeError(
            "Gemini API key not configured. Set GEMINI_API_KEY in .env or the environment, or pass --api-key."
        )
    return context_urls


def run_generation(config: GenerationConfig) -> None:
    """Run the complete resumable generation lifecycle for one config."""
    context_urls = validate_generation_config(config)

    source_title = core.derive_source_title(config.video_path)
    manifest, chunk_dir = build_manifest(config)
    os.makedirs(chunk_dir, exist_ok=True)
    lock_file = None
    staging_vtt = None
    completed = False

    try:
        lock_file = acquire_lock(chunk_dir)
        print(f"Using work directory: {chunk_dir}")

        if config.audio_refine:
            if not media.has_audio_stream(config.video_path):
                raise RuntimeError(
                    "Failed to extract complete audio. "
                    "The source may not contain an audio stream."
                )
            audio_path, audio_duration, _source_duration, _cached = (
                media.extract_complete_audio(config.video_path, chunk_dir)
            )

        # 1. Split video.
        media.split_video(str(config.video_path), chunk_dir, config.chunk_dur, manifest)

        chunks = media.list_chunks(chunk_dir)
        if not chunks:
            raise RuntimeError("No video chunks were created")

        # 2. Process chunks concurrently through one API worker pool.
        failed = process_chunks(
            config.api_key,
            config.base_url,
            chunk_dir,
            chunks,
            config.workers,
            config.model,
            manifest["chunk_mime"],
            config.chunk_thinking_level,
            source_title,
        )
        if failed:
            raise RuntimeError(
                f"Failed to process {len(failed)} chunk(s): {', '.join(sorted(failed))}. "
                f"Keeping {chunk_dir} so you can retry."
            )

        # 3. Stitch chunks at actual segment offsets into a work artifact.
        stitched_vtt = Path(chunk_dir) / "stitched.vtt"
        stitch(chunk_dir, stitched_vtt)
        current_vtt = stitched_vtt

        # 4. Boundary-limited audio refinement over the complete audio.
        if config.audio_refine:
            audio_refined_vtt = Path(chunk_dir) / "audio_refined.vtt"
            gemini.boundary_audio_refine_subtitles(
                stitched_vtt=stitched_vtt,
                audio_path=audio_path,
                audio_duration=audio_duration,
                boundaries=[chunk["start"] for chunk in chunks[1:]],
                work_dir=chunk_dir,
                output_vtt=audio_refined_vtt,
                api_key=config.api_key,
                base_url=config.base_url,
                model_name=config.audio_refine_model
                or gemini.DEFAULT_AUDIO_REFINE_MODEL,
                source_title=source_title,
            )
            current_vtt = audio_refined_vtt

        # 5. Global text refinement or direct atomic publication.
        if config.refine_text:
            output_path = config.output_path
            fd, staging_name = tempfile.mkstemp(
                prefix=f".{output_path.name}.",
                suffix=".staging.vtt",
                dir=output_path.parent,
            )
            staging_vtt = Path(staging_name)
            os.close(fd)
            gemini.global_refine_subtitles(
                input_vtt=str(current_vtt),
                output_vtt=str(staging_vtt),
                api_key=config.api_key,
                base_url=config.base_url,
                model_name=config.refine_model or config.model,
                thinking_level=gemini.REFINEMENT_THINKING_LEVEL,
                source_title=source_title,
                context_urls=context_urls,
            )
            os.replace(staging_vtt, config.output_path)
        else:
            published = core.canonicalize_speaker_casing(webvtt.read(str(current_vtt)))
            io.atomic_save_vtt(published, config.output_path)

        completed = True

    finally:
        try:
            if staging_vtt is not None:
                staging_vtt.unlink(missing_ok=True)

            # 6. Cleanup.
            if completed and os.path.exists(chunk_dir):
                print(f"Cleaning up temporary directory: {chunk_dir}")
                clean_completed_work(chunk_dir)
        finally:
            release_lock(lock_file)

File: internal/pipeline/pipeline.go
package pipeline

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"syscall"

	"github.com/gotenksIN/video-subtitler/internal/core"
	"github.com/gotenksIN/video-subtitler/internal/gemini"
	"github.com/gotenksIN/video-subtitler/internal/media"
	"github.com/gotenksIN/video-subtitler/internal/storage"
	"github.com/gotenksIN/video-subtitler/internal/vtt"
)

const ChunkRoot = "temp_video_chunks"
const DefaultWorkers = 7

type Config struct {
	VideoPath        string
	OutputPath       string
	Model            string
	APIKey           string
	BaseURL          string
	RefineModel      string
	AudioRefineModel string
	ThinkingLevel    string
	ChunkDur         int
	Workers          int
	AudioRefine      bool
	RefineText       bool
	ContextURLs      []string
}

func (config Config) ChunkThinking() string {
	if config.ThinkingLevel == "" {
		return gemini.DefaultThinkingLevel
	}
	return config.ThinkingLevel
}
func BuildManifest(config Config) (map[string]any, string, error) {
	extension, mimeType, codec, err := media.ProbeVideoFormat(config.VideoPath)
	if err != nil {
		return nil, "", err
	}
	fingerprint, err := storage.Fingerprint(config.VideoPath)
	if err != nil {
		return nil, "", err
	}
	manifest := map[string]any{
		"video":                fingerprint,
		"chunk_dur":            config.ChunkDur,
		"format":               "stream-copy-v1",
		"mode":                 "generate",
		"model":                config.Model,
		"chunk_thinking_level": config.ChunkThinking(),
		"chunk_ext":            extension,
		"chunk_mime":           mimeType,
		"video_codec":          codec,
	}
	hash, err := storage.HashJSON(manifest)
	if err != nil {
		return nil, "", err
	}
	return manifest, filepath.Join(ChunkRoot, hash[:16]), nil
}

type Lock struct {
	file *os.File
}

func AcquireLock(dir string) (*Lock, error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, err
	}
	return acquireFileLock(filepath.Join(dir, ".lock"), dir)
}

func AcquireOutputLock(outputPath string) (*Lock, error) {
	absolutePath, err := filepath.Abs(outputPath)
	if err != nil {
		return nil, err
	}
	lockPath := filepath.Join(
		filepath.Dir(absolutePath),
		"."+filepath.Base(absolutePath)+".video-subtitler.lock",
	)
	return acquireFileLock(lockPath, absolutePath)
}

func acquireFileLock(path, resource string) (*Lock, error) {
	file, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0o644)
	if err != nil {
		return nil, err
	}
	if err = syscall.Flock(int(file.Fd()), syscall.LOCK_EX|syscall.LOCK_NB); err != nil {
		contents, readErr := os.ReadFile(path)
		closeErr := file.Close()
		if readErr != nil {
			return nil, readErr
		}
		if closeErr != nil {
			return nil, closeErr
		}
		pid := strings.TrimSpace(string(contents))
		detail := ""
		if _, parseErr := strconv.Atoi(pid); parseErr == nil {
			detail = " (PID " + pid + ")"
		}
		return nil, fmt.Errorf("another run%s is already using %s", detail, resource)
	}
	if err := file.Truncate(0); err != nil {
		_ = file.Close()
		return nil, err
	}
	if _, err := file.Seek(0, 0); err != nil {
		_ = file.Close()
		return nil, err
	}
	if _, err := fmt.Fprint(file, os.Getpid()); err != nil {
		_ = file.Close()
		return nil, err
	}
	if err := file.Sync(); err != nil {
		_ = file.Close()
		return nil, err
	}
	return &Lock{file: file}, nil
}
func (lock *Lock) Close() {
	if lock != nil && lock.file != nil {
		_ = syscall.Flock(int(lock.file.Fd()), syscall.LOCK_UN)
		_ = lock.file.Close()
	}
}
func CleanCompleted(dir string) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		if entry.Name() == ".lock" {
			continue
		}
		if err = os.RemoveAll(filepath.Join(dir, entry.Name())); err != nil {
			return err
		}
	}
	return nil
}
func Validate(config Config) ([]string, error) {
	contextURLs, err := core.ValidateContextURLs(config.ContextURLs)
	if err != nil {
		return nil, err
	}
	if config.ChunkDur <= 0 {
		return nil, fmt.Errorf("--chunk-dur must be greater than 0")
	}
	if config.Workers <= 0 {
		return nil, fmt.Errorf("--workers must be greater than 0")
	}
	if err = gemini.ValidateThinking(config.Model, config.ChunkThinking()); err != nil {
		return nil, err
	}
	if _, err = os.Stat(config.VideoPath); err != nil {
		return nil, fmt.Errorf("video file not found: %s", config.VideoPath)
	}
	videoPath, err := filepath.Abs(config.VideoPath)
	if err != nil {
		return nil, err
	}
	outputPath, err := filepath.Abs(config.OutputPath)
	if err != nil {
		return nil, err
	}
	if resolved, resolveErr := filepath.EvalSymlinks(videoPath); resolveErr == nil {
		videoPath = resolved
	}
	if resolved, resolveErr := filepath.EvalSymlinks(outputPath); resolveErr == nil {
		outputPath = resolved
	}
	if videoPath == outputPath {
		return nil, fmt.Errorf("--output must not resolve to the source video")
	}
	if config.APIKey == "" {
		return nil, fmt.Errorf("Gemini API key not configured; set GEMINI_API_KEY in .env or the environment, or pass --api-key")
	}
	return contextURLs, nil
}
func ProcessChunks(ctx context.Context, config Config, dir string, chunks []media.Chunk, mime, title string, names []string) []string {
	fmt.Printf("Processing %d chunks using %d workers...\n", len(chunks), config.Workers)
	jobs := make(chan media.Chunk)
	failed := make(chan string, len(chunks))
	var workers sync.WaitGroup
	for range config.Workers {
		workers.Add(1)
		go func() {
			defer workers.Done()
			for chunk := range jobs {
				succeeded := gemini.ProcessChunk(
					ctx,
					config.APIKey,
					config.BaseURL,
					dir,
					chunk,
					config.Model,
					mime,
					config.ChunkThinking(),
					title,
					names,
				)
				if !succeeded {
					failed <- chunk.Name
				}
			}
		}()
	}
	for _, chunk := range chunks {
		jobs <- chunk
	}
	close(jobs)
	workers.Wait()
	close(failed)
	var failedNames []string
	for name := range failed {
		failedNames = append(failedNames, name)
	}
	sort.Strings(failedNames)
	return failedNames
}

type stitchedEntry struct {
	start      float64
	end        float64
	text       string
	chunkIndex int
}

func Stitch(dir, output string) error {
	fmt.Println("Stitching chunks into final VTT...")
	chunks := media.ListChunks(dir)
	chunksByIndex := map[int]media.Chunk{}
	var boundaryStarts []float64
	for index, chunk := range chunks {
		chunksByIndex[chunk.Idx] = chunk
		if index > 0 {
			boundaryStarts = append(boundaryStarts, chunk.Start)
		}
	}
	resultNames, err := subtitleResultNames(dir)
	if err != nil {
		return err
	}
	if len(resultNames) != len(chunks) {
		return fmt.Errorf("invalid subtitle results: expected %d chunks, found %d", len(chunks), len(resultNames))
	}
	var entries []stitchedEntry
	for _, name := range resultNames {
		chunkIndex, err := subtitleResultIndex(name)
		if err != nil {
			return err
		}
		chunk, ok := chunksByIndex[chunkIndex]
		if !ok {
			return fmt.Errorf("invalid subtitle results: unexpected chunk index %d", chunkIndex)
		}
		captions, err := readCaptionResult(filepath.Join(dir, name))
		if err != nil {
			return err
		}
		for _, caption := range captions {
			start, err := core.ParseTime(caption.Start)
			if err != nil {
				return fmt.Errorf("invalid caption timing in %s: %w", name, err)
			}
			end, err := core.ParseTime(caption.End)
			if err != nil || end <= start {
				return fmt.Errorf("invalid caption timing in %s", name)
			}
			entries = append(entries, stitchedEntry{
				start:      chunk.Start + start,
				end:        chunk.Start + end,
				text:       caption.Text,
				chunkIndex: chunkIndex,
			})
		}
	}
	sort.SliceStable(entries, func(i, j int) bool { return entries[i].start < entries[j].start })
	entries = mergeVisualBoundaryEntries(entries, boundaryStarts)
	outputFile := &vtt.File{}
	for _, entry := range entries {
		outputFile.Cues = append(outputFile.Cues, vtt.Cue{
			Start: core.MustFormatTime(entry.start),
			End:   core.MustFormatTime(entry.end),
			Text:  entry.text,
		})
	}
	if err := outputFile.SaveAtomic(output); err != nil {
		return err
	}
	fmt.Printf("Successfully saved to %s with %d total captions.\n", output, len(outputFile.Cues))
	return nil
}

func subtitleResultNames(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var names []string
	for _, entry := range entries {
		name := entry.Name()
		if strings.HasPrefix(name, "subtitle_chunk_") && strings.HasSuffix(name, ".json") {
			names = append(names, name)
		}
	}
	sort.Strings(names)
	return names, nil
}

func subtitleResultIndex(name string) (int, error) {
	value := strings.TrimSuffix(strings.TrimPrefix(name, "subtitle_chunk_"), ".json")
	return strconv.Atoi(value)
}

func readCaptionResult(path string) ([]core.Caption, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var captions []core.Caption
	if err := json.Unmarshal(data, &captions); err != nil {
		return nil, fmt.Errorf("invalid subtitle result %s: %w", filepath.Base(path), err)
	}
	return captions, nil
}

func mergeVisualBoundaryEntries(entries []stitchedEntry, boundaries []float64) []stitchedEntry {
	for {
		merged := false
		for currentIndex, current := range entries {
			if current.chunkIndex >= len(boundaries) {
				continue
			}
			boundary := boundaries[current.chunkIndex]
			for followingIndex := currentIndex + 1; followingIndex < len(entries); followingIndex++ {
				following := entries[followingIndex]
				if !canMergeVisualEntries(current, following, boundary) {
					continue
				}
				entries[currentIndex] = stitchedEntry{
					start:      current.start,
					end:        following.end,
					text:       strings.TrimSpace(current.text),
					chunkIndex: following.chunkIndex,
				}
				entries = append(entries[:followingIndex], entries[followingIndex+1:]...)
				merged = true
				break
			}
			if merged {
				break
			}
		}
		if !merged {
			return entries
		}
	}
}

func canMergeVisualEntries(current, following stitchedEntry, boundary float64) bool {
	return following.chunkIndex == current.chunkIndex+1 &&
		core.ClassifyCueText(current.text) == "editorial" &&
		core.ClassifyCueText(following.text) == "editorial" &&
		strings.TrimSpace(current.text) == strings.TrimSpace(following.text) &&
		abs(boundary-current.end) <= 0.5 &&
		abs(following.start-boundary) <= 0.5
}
func abs(value float64) float64 {
	if value < 0 {
		return -value
	}
	return value
}
func Run(ctx context.Context, config Config) error {
	contextURLs, err := Validate(config)
	if err != nil {
		return err
	}
	title := core.DeriveSourceTitle(config.VideoPath)
	manifest, workDir, err := BuildManifest(config)
	if err != nil {
		return err
	}
	if err = os.MkdirAll(workDir, 0o755); err != nil {
		return err
	}
	workLock, err := AcquireLock(workDir)
	if err != nil {
		return err
	}
	defer workLock.Close()
	outputLock, err := AcquireOutputLock(config.OutputPath)
	if err != nil {
		return err
	}
	defer outputLock.Close()
	audioPath, audioDuration, err := prepareAudio(config, workDir)
	if err != nil {
		return err
	}
	preflight, err := preparePreflight(ctx, config, workDir, title, contextURLs)
	if err != nil {
		return err
	}
	if err = media.SplitVideo(config.VideoPath, workDir, config.ChunkDur, manifest); err != nil {
		return err
	}
	chunks := media.ListChunks(workDir)
	if len(chunks) == 0 {
		return fmt.Errorf("no video chunks were created")
	}
	failed := ProcessChunks(
		ctx,
		config,
		workDir,
		chunks,
		manifest["chunk_mime"].(string),
		title,
		preflight.GroundedNames,
	)
	if len(failed) > 0 {
		return fmt.Errorf(
			"failed to process %d chunk(s): %s; keeping %s so you can retry",
			len(failed),
			strings.Join(failed, ", "),
			workDir,
		)
	}
	stitchedPath := filepath.Join(workDir, "stitched.vtt")
	if err = Stitch(workDir, stitchedPath); err != nil {
		return err
	}
	currentPath := stitchedPath
	if config.AudioRefine {
		currentPath = filepath.Join(workDir, "audio_refined.vtt")
		boundaries := []float64{}
		for _, chunk := range chunks[1:] {
			boundaries = append(boundaries, chunk.Start)
		}
		if err = gemini.BoundaryAudioRefine(
			ctx,
			stitchedPath,
			audioPath,
			workDir,
			currentPath,
			config.APIKey,
			config.BaseURL,
			first(config.AudioRefineModel, gemini.DefaultAudioRefineModel),
			title,
			audioDuration,
			boundaries,
		); err != nil {
			return err
		}
	}
	if err = publishResult(ctx, config, currentPath, title, contextURLs, preflight); err != nil {
		return err
	}
	fmt.Printf("Cleaning up temporary directory: %s\n", workDir)
	return CleanCompleted(workDir)
}

func prepareAudio(config Config, workDir string) (string, float64, error) {
	if !config.AudioRefine {
		return "", 0, nil
	}
	if !media.HasAudio(config.VideoPath) {
		return "", 0, fmt.Errorf("failed to extract complete audio; the source may not contain an audio stream")
	}
	audioPath, duration, _, _, err := media.ExtractAudio(config.VideoPath, workDir)
	return audioPath, duration, err
}

func preparePreflight(
	ctx context.Context,
	config Config,
	workDir, title string,
	contextURLs []string,
) (core.PreflightContext, error) {
	if cached, ok := gemini.LoadPreflight(workDir); ok {
		return cached, nil
	}
	preflight, err := gemini.RunPreflight(
		ctx,
		config.APIKey,
		config.BaseURL,
		first(config.RefineModel, config.Model),
		"high",
		title,
		contextURLs,
	)
	if err != nil {
		return core.PreflightContext{}, err
	}
	if err := gemini.StorePreflight(workDir, preflight); err != nil {
		return core.PreflightContext{}, err
	}
	return preflight, nil
}

func publishResult(
	ctx context.Context,
	config Config,
	currentPath, title string,
	contextURLs []string,
	preflight core.PreflightContext,
) error {
	if !config.RefineText {
		outputFile, err := vtt.Read(currentPath)
		if err != nil {
			return err
		}
		outputFile.Cues = core.CanonicalizeSpeakerCasing(outputFile.Cues, nil)
		return outputFile.SaveAtomic(config.OutputPath)
	}
	outputDir := filepath.Dir(config.OutputPath)
	temporaryFile, err := os.CreateTemp(
		outputDir,
		"."+filepath.Base(config.OutputPath)+".*.staging.vtt",
	)
	if err != nil {
		return err
	}
	stagingPath := temporaryFile.Name()
	if err := temporaryFile.Close(); err != nil {
		_ = os.Remove(stagingPath)
		return err
	}
	defer os.Remove(stagingPath)
	if err := gemini.Refine(
		ctx,
		currentPath,
		stagingPath,
		config.APIKey,
		config.BaseURL,
		first(config.RefineModel, config.Model),
		"high",
		title,
		contextURLs,
		nil,
		&preflight,
	); err != nil {
		return err
	}
	return os.Rename(stagingPath, config.OutputPath)
}
func first(preferred, fallback string) string {
	if preferred != "" {
		return preferred
	}
	return fallback
}
